log-elasticsearch/src/Log/Backend/ElasticSearch/V5.hs

```diff
@@ -166,131 +128,88 @@ elasticSearchLogger esConf@ElasticSearchConfig{..} genRandomWord = do
-          void $ bulk (V.map (toBulk index baseID) newMsgs))
-    (elasticSearchSync indexRef)
+                      ) (modifyData Nothing v) acc
+                in adjustFailedMessagesWith modifyData jsonMsgs responses
+          -- Attempt to put modified messages.
+          newReply <- responseBody <$> bulkIndex version env esConf index newMsgs
+          case checkForBulkErrors newMsgs newReply of
+            Nothing -> printEsError "unexpected response" newReply
+            Just (newHasErrors, newResponses) -> when newHasErrors $ do
+              -- If some of the messages failed again (it might happen e.g. if
+              -- data contains an array with elements of different types), drop
+              -- their data field.
+              let newerMsgs =
+                    let modifyData :: Maybe Value -> Value -> Value
+                        modifyData (Just err) Object{} = object [ "__es_error" .= err ]
+                        modifyData _ v = v
+                    in adjustFailedMessagesWith modifyData newMsgs newResponses
+              -- Ignore any further errors.
+              void $ bulkIndex version env esConf index newerMsgs)
+    (refreshIndex env =<< readIORef indexRef)
   where
-    mapping = MappingName esMapping
-
-    elasticSearchSync :: IORef IndexName -> IO ()
-    elasticSearchSync indexRef = do
-      indexName <- readIORef indexRef
-      void . runBH_ esConf $ refreshIndex indexName
-
-    retryOnException :: forall r. IO r -> IO r
-    retryOnException m = try m >>= \case
+    -- Process reply of bulk indexing to get responses for each index operation
+    -- and check whether any insertion failed.
+    checkForBulkErrors
+      :: V.Vector a
+      -> Value
+      -> Maybe (Bool, V.Vector Object)
+    checkForBulkErrors jsonMsgs replyBody = do
+      Object response <- pure replyBody
+      Bool hasErrors  <- "errors" `H.lookup` response
+      Array jsonItems <- "items"  `H.lookup` response
+      items <- F.forM jsonItems $ \v -> do
+        Object item   <- pure v
+        Object index_ <- "index" `H.lookup` item
+          -- ES <= 2.x returns 'create' for some reason, so consider both.
+          <|> "create" `H.lookup` item
+        pure index_
+      guard $ V.length items == V.length jsonMsgs
+      pure (hasErrors, items)
+
+    adjustFailedMessagesWith
+      :: (Maybe err -> obj -> obj)
+      -> V.Vector (H.HashMap T.Text obj)
+      -> V.Vector (H.HashMap T.Text err)
+      -> V.Vector (H.HashMap T.Text obj)
+    adjustFailedMessagesWith f jsonMsgs responses =
+      let failed = V.imapMaybe (\n item -> (n, ) <$> "error" `H.lookup` item) responses
+      in (`V.map` failed) $ \(n, err) -> H.adjust (f $ Just err) "data" $ jsonMsgs V.! n
+
+    printEsError msg body =
+      T.putStrLn $ "elasticSearchLogger: " <> msg <> " " <> prettyJson body
+
+    retryOnException :: forall r. IORef (Maybe EsVersion) -> IO r -> IO r
```

Comment:
We can discuss it, but we already have bounded message queue in place for longer ES outages.

I'd say it's better for the code to keep trying indefinitely than you having to restart your app after 10 minutes because the logger died.

Reply:
Ok, sure, I think you're right. I just thought about eid dev/testbed environment where this thrown all the time indefinitely, because ES isn't configured, but in that case it should be solved differently anyway.